src/vsr.zig

```diff
@@ -1174,27 +1311,85 @@ test "Headers.ViewChangeSlice.view_for_op" {
 
 /// The headers of a SV or DVC message.
 const ViewChangeHeadersArray = struct {
+    command: ViewChangeCommand,
     array: Headers.Array,
 
     pub fn root(cluster: u32) ViewChangeHeadersArray {
-        var array = Headers.Array{ .buffer = undefined };
-        array.appendAssumeCapacity(Header.root_prepare(cluster));
-        return ViewChangeHeadersArray.init(array);
+        return ViewChangeHeadersArray.init_from_slice(.start_view, &.{
+            Header.root_prepare(cluster),
+        });
+    }
+
+    pub fn init_from_slice(
+        command: ViewChangeCommand,
+        slice: []const Header,
+    ) ViewChangeHeadersArray {
+        const headers = ViewChangeHeadersArray{
+            .command = command,
+            .array = Headers.Array.fromSlice(slice) catch unreachable,
+        };
+        headers.verify();
+        return headers;
+    }
+
+    fn init_from_array(command: ViewChangeCommand, array: Headers.Array) ViewChangeHeadersArray {
+        const headers = ViewChangeHeadersArray{
+            .command = command,
+            .array = array,
+        };
+        headers.verify();
+        return headers;
     }
 
-    pub fn init_from_slice(slice: []const Header) ViewChangeHeadersArray {
-        Headers.ViewChangeSlice.verify(slice);
-        return .{ .array = Headers.Array.fromSlice(slice) catch unreachable };
+    pub fn verify(headers: *const ViewChangeHeadersArray) void {
+        (ViewChangeHeadersSlice{
+            .command = headers.command,
+            .slice = headers.array.constSlice(),
+        }).verify();
     }
 
-    pub fn replace(headers: *ViewChangeHeadersArray, slice: []const Header) void {
-        Headers.ViewChangeSlice.verify(slice);
+    pub fn start_view_into_do_view_change(headers: *ViewChangeHeadersArray) void {
+        assert(headers.command == .start_view);
+        headers.command = .do_view_change;
+
+        // Remove the "hook" checkpoint trigger(s), since they would create gaps in the ops.
+        // (There are at most 2 hooks).
+        var removed: usize = 0;
+        while (headers.array.get(headers.array.len - 2).op !=
```

Comment:
Hm, do we always have at least to?

```
        if (headers.command == .start_view) {
            assert(headers.slice.len >= @minimum(
                constants.view_change_headers_suffix_max,
                op_head + 1, // +1 to include the head itself
            ));
        }
```

makes me thing that one is possible. If it isn't, that should also assert that `op_head >= 1`

Reply:
`start_view.headers.len == 1` possible in general (and in `verify()`), but not in `start_view_into_do_view_change`.

Adding this explanation:

```zig
    pub fn start_view_into_do_view_change(headers: *ViewChangeHeadersArray) void {
        assert(headers.command == .start_view);
        // This function is only called by a replica that is lagging behind the primary's
        // checkpoint, so the start_view has a full suffix of headers.
        assert(headers.array.get(0).op > constants.journal_slot_count);
        assert(headers.array.len >= constants.view_change_headers_suffix_max);
```

and this postcondition after the hooks are removed:

```zig
        assert(headers.array.len == constants.view_change_headers_suffix_max);
```